content/en/docs/reference/setup-tools/kubeadm/kubeadm-init.md

```diff
@@ -241,6 +218,20 @@ or `kubeadm upgrade apply`), kubeadm respects the value of `UnversionedKubeletCo
 (during `kubeadm join`, `kubeadm reset`, `kubeadm upgrade ...`), kubeadm attempts to use unversioned ConfigMap name first;
 if that does not succeed, kubeadm falls back to using the legacy (versioned) name for that ConfigMap.
 
+`UpgradeAddonsBeforeControlPlane`
+: This is as a **disabled** feature gate that was introduced for Kubernetes v1.28, in order to allow reactivating a legacy
+and deprecated behavior during cluster upgrade. For kubeadm versions prior to v1.28, kubeadm upgrades cluster addons (including
+CoreDNS and kube-proxy) immediately during `kubeadm upgrade apply`, regardless of whether there are other control plane
+instances that have not been upgraded. This may cause compatibility problems. Since v1.28, kubeadm defaults to a mode that
+always checks whether all the control plane instances have been upgraded before starting to upgrade the addons. This behavior
+is applied to both `kubeadm upgrade apply` and `kubeadm upgrade node`. kubeadm determines whether a control plane instance
+has been upgraded by checking whether the image of the kube-apiserver Pod has been upgraded. You must perform control plane
+instances upgrade sequentially or at least ensure that the last control plane instance upgrade is not started until all the
+other control plane instances have been upgraded completely, and the addons upgrade will be performed after the last control plane
+instance is upgraded. The deprecated `UpgradeAddonsBeforeControlPlane` feature gate gives you a chance to keep the old upgrade
+behavior. You should not need this old behavior; if you do, you should consider changing your cluster or upgrade processes. 
```

Comment:
Ideally, add all this text in to the _main_ branch (which at time of writing documents v1.30), and then once it has merged, update a couple of lines to make it correct for v1.31

Reply:
thanks @sftim 
Does that mean we can commit this to the main branch first? Will that affect the documentation content of v1.30? We expect this change to only be seen in v1.31 and later.

Reply:
On the contrary, my take of this change is that the gate can be used on v1.28-v1.30.
It is something that we should have documented since v1.28 and then **removed** in v1.31.

How about we document this into v1.30 (main) branch and then backport to v1.28-v1.29. The update we then need for v1.31 would be **deleting** or **drastically simplify** this paragraph because the gate has been removed.

Reply:
the gate existed as "deprecated" since 1.28 up until 1.30 under "List of deprecated feature gates:". it started as a deprecated FG. see https://v1-28.docs.kubernetes.io/docs/reference/setup-tools/kubeadm/kubeadm-init/#feature-gates
i don't understand why this change should be backported or targeting anything other than dev-1.31.

>  for v1.31 would be deleting or drastically simplify this paragraph because the gate has been removed.

this can be done, but we haven't done such simplifications for prior removals of FGs.

Reply:
A gate, even for one that is deprecated when first introduced, is still something a user can use. I misunderstood the whole changes proposed in this PR. Actually this PR is about **moving** rather than **removing** the description about the gate. That should be fine. I mean, we don't need to backport anything.

For v1.31, the "drastically simplified version" could be something like:

> This gate has been removed. It was introduced in v1.28 as a deprecated feature and then removed in v1.31. For documentation on older versions, please switch to the corresponding website version.

I mean, we don't need to add more details for such a "born to be deprecated" feature in v1.31.

Reply:
>> This gate has been removed. It was introduced in v1.28 as a deprecated feature and then removed in v1.31. For documentation on older versions, please switch to the corresponding website version.

> I mean, we don't need to add more details for such a "born to be deprecated" feature in v1.31.

SGTM

Reply:
thanks, updated. @neolit123 @tengqm